vortex-ffi/src/lib.rs

```diff
@@ -41,3 +46,134 @@ pub(crate) unsafe fn to_string_vec(ptr: *const *const c_char, len: c_int) -> Vec
+            #[doc = r" Must be released with [`" $ffi_ident "_free`]."]
+            #[unsafe(no_mangle)]
+            pub unsafe extern "C-unwind" fn [<$ffi_ident _clone>](ptr: *const $ffi_ident) -> *const $ffi_ident {
+                if ptr.is_null() {
+                    vortex::error::vortex_panic!("null pointer");
+                }
+                $ffi_ident::new($ffi_ident::into_arc(ptr.cast_mut()).clone())
+            }
+
+            #[doc = r" Free an owned [`" $ffi_ident "`] object."]
+            #[unsafe(no_mangle)]
+            pub unsafe extern "C-unwind" fn [<$ffi_ident _free>](ptr: *const $ffi_ident) {
+                if ptr.is_null() {
+                    vortex::error::vortex_panic!("null pointer");
+                }
+                drop($ffi_ident::into_arc(ptr.cast_mut()))
+            }
+        }
+    };
+}
+
+/// Define a native FFI type that uses an [`std::sync::Arc`] wrapper.
+#[macro_export]
+macro_rules! arc_wrapper {
+    ($(#[$meta:meta])* $T:ty, $ffi_ident:ident) => {
+        paste::paste! {
+            $(#[$meta])*
+            #[allow(non_camel_case_types)]
+            pub struct $ffi_ident {
+                inner: $T
+            }
+
+            #[allow(dead_code)]
+            impl $ffi_ident {
+                /// Wrap an owned object into a raw pointer.
+                pub(crate) fn new(obj: std::sync::Arc<$T>) -> *const $ffi_ident {
+                    std::sync::Arc::into_raw(obj).cast()
+                }
+
+                /// Wrap a borrowed object into a raw pointer.
+                pub(crate) fn new_ref(obj: &$T) -> *const $ffi_ident {
+                    obj as *const $T as *const $ffi_ident
+                }
+
+                /// Extract a borrowed reference from a const pointer.
+                pub(crate) fn as_ref<'a>(ptr: *const $ffi_ident) -> &'a $T {
+                    use vortex::error::VortexExpect;
+                    &unsafe { ptr.as_ref() }
+                        .vortex_expect("null pointer")
+                        .inner
+                }
+
+                /// Extract an owned reference.
+                pub(crate) fn into_arc(ptr: *const $ffi_ident) -> std::sync::Arc<$T>{
+                    if ptr.is_null() {
+                        vortex::error::vortex_panic!("null pointer");
+                    }
+                    unsafe { std::sync::Arc::from_raw(ptr.cast::<$T>()) }
+                }
+            }
+
+            #[doc = r" Clone a borrowed [`" $ffi_ident "`], returning an owned [`" $ffi_ident "`].\n\n"]
+            #[doc = r" Must be released with [`" $ffi_ident "_free`]."]
+            #[unsafe(no_mangle)]
+            pub unsafe extern "C-unwind" fn [<$ffi_ident _clone>](ptr: *const $ffi_ident) -> *mut $ffi_ident {
+                if ptr.is_null() {
+                    vortex::error::vortex_panic!("null pointer");
+                }
+                unsafe { std::sync::Arc::increment_strong_count(ptr) };
+                ptr.cast_mut()
```

Comment:
I don't think this is right? if you increment the strong count of the Arc, that doesn't mean you can just treat it as a unique pointer now

Reply:
Yeah, this should return `*const ffi_ident`, but it's correct that it's now "owned"